test/Test/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -314,3 +318,175 @@ prop_single_entry (Ref.PageContentSingle k op) =
+        , if null pKeys
+          then "<NONE>"
+          else  show $ maximum pKeys < key
+        ]
+
+-- |
+-- Helper data-type for generating arbitrary pages and offset keys
+-- with a useful distribution of hit and misses for testing purposes.
+data RawPageAndOffset = RawPageAndOffset RawPage SerialisedKey
+  deriving stock (Eq, Show)
+
+instance Arbitrary RawPageAndOffset where
+
+  arbitrary = do
+    Ref.PageContentOrdered kops <- arbitrary
+    let rawpage = fst . Ref.toRawPage $ Ref.PageContentFits kops
+    RawPageAndOffset rawpage <$> case kops of
+      -- If there are no kOps, then generate any offset key
+      [] -> Ref.toSerialisedKey <$> arbitrary
+
+      -- Otherwise, if at least one kOp exists then generate the offset key according to
+      -- the following probability distribution which is conditional on kOps:
+      --   * (1/2): Offset Key exists within the page
+      --   * (1/6): Offset Key goes before the page
+      --   * (1/6): Offset Key goes after the page
+      --   * (1/6): Offset Key missing within the page
+      --            (Assuming that there are two or more keys)
+      xs ->
+        let keys = fst <$> xs
+            keySmall = Ref.toSerialisedKey $ minimum keys
+            keyLarge = Ref.toSerialisedKey $ maximum keys
+
+            missingWithinRange
+              | keySmall == keyLarge = []
+              | otherwise =
+                let missingKey = do
+                      let n = length keys - 1
+                      afterIndex <- (`mod` n) <$> arbitrary
+                      let loKey = Ref.toSerialisedKey $ keys !! afterIndex
+                          hiKey = Ref.toSerialisedKey $ keys !! (afterIndex + 1)
+                      arbitrarySerialisedKeyBetween loKey hiKey
+                in  [ (1, missingKey) ]
+
+        in  frequency $
+              [ (3, Ref.toSerialisedKey <$> elements keys)
+              , (1, arbitrarySerialisedKeyLessThan keySmall)
+              , (1, arbitrarySerialisedKeyMoreThan keyLarge)
+              ] <> missingWithinRange
+
+-- |
+-- Generate an arbitrary 'SerialisedKey' between the given two keys.
+--
+-- The first key is assumed to be the "smaller" of the pair.
+--
+-- NOTE: Under pathological input conditions, there is an infintesimally small chance
+-- that this will generate /the same/ key as the larger value. This will not effect the
+-- validity of 'Arbitrary' instances which depend on this function.
+arbitrarySerialisedKeyBetween :: SerialisedKey -> SerialisedKey -> Gen SerialisedKey
+arbitrarySerialisedKeyBetween (SerialisedKey (RawBytes loBytes)) (SerialisedKey (RawBytes hiBytes)) =
+    SerialisedKey . RawBytes <$> VP.zipWithM withinByte loBytes hiBytes
+  where
+    withinByte !lhs !rhs
+      | lhs == rhs = pure rhs
+      | otherwise =
+        let (!lo, !hi)
+              | lhs < rhs = (lhs, rhs)
+              | otherwise = (rhs, lhs)
+            !diff = hi - lo
+        in  (succ lo +) . (`mod` diff) <$> arbitrary
```

Comment:
This does a slightly weird thing, where *each byte* is between the corresponding bytes of the two keys. This limits the set of keys that get generated quite a bit. If the first byte is already smaller/bigger than the first byte of the lower/upper bound, then the remaining bytes could be completely arbitrary.

Reply:
We can leave this as is, since the only thing observed about keys is their ordering towards each other. Still feels a bit off.

A potential solution would be to use a simpler generator (maybe only 50:50 existing key or arbitrary key), with labels for the cases specifically targetted here, to make sure the distribution looks good. Recognizing a property of the input is usually simpler than having to specifically generate input with that property.

Reply:
I took your advice to just simplify the generator and change the distribution to choose an existing key with $\mathbf{P}\left\[\frac{1}{2}\right\]$ and an `arbitrary` key otherwise.

Reply:
Actually, I've reversed my position in light of new information. I annotated the tests with [`label`](https://hackage.haskell.org/package/tasty-quickcheck/docs/Test-Tasty-QuickCheck.html#v:label) to see how the 50/50 distribution of random/existing key selection panned out. There was *essentially* no coverage of the "before page" and "after page" cases. So I decided to bring back the computationally more complicated, but more predictable distribution of outcomes.

After bringing back the original, more complex distribution, I updated the sub-generators to have an expanded value range. The updated ranges address the previously "weird" location-independent byte generation which you had commented on. 

I had chosen that admittedly weird generation technique to simplify the generation process. To correctly generate any key within an upper and/or lower byte-stream, I would need to track if I had generated a smaller/larger byte earlier in the sequence. This required dragging state through the QuickCheck generator which I had lazily avoided doing. That state dragging has been added now.

Reply:
The usage of `min`/`max` still doesn't seem 100% right, but it should be good enough and lead to a good distribution of test cases.